avro-serializer/src/main/java/io/confluent/kafka/serializers/AbstractKafkaAvroSerDe.java

```diff
@@ -80,26 +84,78 @@ protected void configureClientProperties(AbstractKafkaAvroSerDeConfig config) {
   /**
    * Get the subject name for the given topic and value type.
    */
-  protected static String getSubjectName(String topic, boolean isKey) {
+  protected String getSubjectName(String topic, boolean isKey, Object value) {
     if (isKey) {
-      return topic + "-key";
+      if (keySubjectNameStrategy.equals("topic-key")) {
+        return topic + "-key";
+      } else if (keySubjectNameStrategy.equals("type")) {
+        return getRecordName(value, isKey);
+      } else if (keySubjectNameStrategy.equals("topic-type")) {
+        return topic + "-" + getRecordName(value, isKey);
+      } else {
+        throw new SerializationException("Unknown value for "
+                + AbstractKafkaAvroSerDeConfig.KEY_SUBJECT_NAME_STRATEGY + ": "
+                + keySubjectNameStrategy);
+      }
     } else {
-      return topic + "-value";
+      if (valueSubjectNameStrategy.equals("topic-value")) {
+        return topic + "-value";
+      } else if (valueSubjectNameStrategy.equals("type")) {
+        return getRecordName(value, isKey);
+      } else if (valueSubjectNameStrategy.equals("topic-type")) {
+        return topic + "-" + getRecordName(value, isKey);
+      } else {
+        throw new SerializationException("Unknown value for "
+                + AbstractKafkaAvroSerDeConfig.VALUE_SUBJECT_NAME_STRATEGY + ": "
+                + valueSubjectNameStrategy);
+      }
```

Comment:
This method is called frequently, and doing all these string comparisons in each call is less than ideal. What do you think about creating a functional interface like:

```
public interface SubjectNamingStrategy {
    String getSubjectName(String topic, Object value);
}
```
with a separate implementation for each strategy (see below). The strategy for keys and values could be instantiated once in `configureClientProperties(...)` above:

```
if ("topic-key".equals(keySubjectNameStrategy)) {
    keySubjectStrategy = new TopicNamingStrategy();
} else if ("type".equals(keySubjectNameStrategy)) {
    keySubjectStrategy = new RecordSchemaNamingStrategy("");
} else if ("topic-type".equals(keySubjectNameStrategy)) {
    keySubjectStrategy = new RecordSchemaNamingStrategy("topic-");
} else {
    throw new SerializationException("Unknown value for "
                + AbstractKafkaAvroSerDeConfig.KEY_SUBJECT_NAME_STRATEGY + ": "
                + keySubjectNameStrategy);
}
if ("topic-value".equals(valueSubjectNameStrategy)) {
    keySubjectStrategy = new TopicNamingStrategy();
} else if ("type".equals(valueSubjectNameStrategy)) {
    valueSubjectStrategy = new RecordSchemaNamingStrategy("");
} else if ("topic-type".equals(valueSubjectNameStrategy)) {
    valueSubjectStrategy = new RecordSchemaNamingStrategy("topic-");
} else {
    throw new SerializationException("Unknown value for "
                + AbstractKafkaAvroSerDeConfig.VALUE_SUBJECT_NAME_STRATEGY + ": "
                + valueSubjectNameStrategy);
}
```

One benefit is that an incorrect value for the configuration property is detected right away. However, the big benefit is that the `getSubjectName` member method that is called frequently by subclasses can be far more efficient by delegating to the correct strategy and forgoing all of the string comparisons:
```
protected String getSubjectName(String topic, boolean isKey, Object value) {
    return isKey ? keyStrategy.getSubjectName(topic, isKey, value) : valueStrategy.getSubjectName(topic, isKey, value);
}
```

Each `SubjectNamingStrategy` implementation would be quite straightforward. For example, the `TopicSubjectNamingStrategy` might be implemented as follows:
```
protected static class TopicNamingStrategy implements SubjectNamingStrategy {
    public String getSubjectName(String topic, boolean isKey, Object value) {
       if (isKey) {
           return topic + "-key";
       }
       return topic + "-value"; 
    }
}
```
while another `RecordSchemaNamingStrategy` implementation could handle both `type` and `topic-type` options by just using different prefixes:
```
protected static class RecordSchemaNamingStrategy implements SubjectNamingStrategy {
    private final String prefix;
    public RecordSchemaNamingStrategy(String prefix) {
        this.prefix = prefix != null ? prefix : "";
    }
    public String getSubjectName(String topic, boolean isKey, Object value) {
        // Null is passed through unserialized, since it has special meaning in
        // log-compacted Kafka topics.
        if (value == null) {
          return null;
        }

        if (value instanceof GenericContainer) {
          Schema schema = ((GenericContainer) value).getSchema();
          if (schema.getType() == Schema.Type.RECORD) {
            return prefix + schema.getFullName();
          }
        }

        // isKey is only used to produce more helpful error messages
        if (isKey) {
          throw new SerializationException("In configuration "
              + AbstractKafkaAvroSerDeConfig.KEY_SUBJECT_NAME_STRATEGY + " = "
              + keySubjectNameStrategy + ", the message key must only be an Avro record");
        } else {
          throw new SerializationException("In configuration "
              + AbstractKafkaAvroSerDeConfig.VALUE_SUBJECT_NAME_STRATEGY + " = "
              + valueSubjectNameStrategy + ", the message value must only be an Avro record");
        }
    }
}
```

Reply:
IMO, I would prefer making the Strategy interface public so that users could choose to use their own topic to subject mapping strategy if they need to.